src/app/clusters/mode-base-server/mode-base-server.h

```diff
@@ -0,0 +1,198 @@
+     */
+    void handleChangeToMode(HandlerContext & ctx, const Commands::ChangeToMode::DecodableType & req);
+
+public:
+    /**
+     * Creates a mode base cluster instance. The Init() function needs to be called for this instance to be registered and
+     * called by the interaction model at the appropriate times.
+     * @param aEndpointId The endpoint on which this cluster exists. This must match the zap configuration.
+     * @param aClusterId The ID of the ModeBase aliased cluster to be instantiated.
+     * @param aFeature The bitmask value that identifies which features are supported by this instance.
+     */
+    Instance(EndpointId aEndpointId, ClusterId aClusterId, uint32_t aFeature) :
+        CommandHandlerInterface(Optional<EndpointId>(aEndpointId), aClusterId),
+        AttributeAccessInterface(Optional<EndpointId>(aEndpointId), aClusterId)
+    {
+        mEndpointId = aEndpointId;
+        mClusterId  = aClusterId;
+        mFeature    = aFeature;
+    }
+
+    ~Instance() override { ModeBaseAliasesInstanceMap.erase(mClusterId); }
+
+    template <typename RequestT, typename FuncT>
+    void HandleCommand(HandlerContext & handlerContext, FuncT func);
+
+    // The following functions should be overridden by the SDK user to implement the business logic of their application.
+
+    /**
+     * This init function will be called during the ModeBase server initialization after the Instance information has been
+     * validated and the Instance has been registered. This can be used to initialise app logic.
+     */
+    virtual CHIP_ERROR AppInit() = 0;
+
+    /**
+     * Returns the number of modes managed by this instance.
+     */
+    virtual uint8_t NumberOfModes();
+
+    /**
+     * Get the mode label of the Nth mode in the list of modes.
+     * @param modeIndex The index of the mode to be returned. It is assumed that modes are indexable from 0 and with no gaps.
+     * @param label a reference to the MutableCharSpan that is to contain the mode label. Use CopyCharSpanToMutableCharSpan
+     * to copy into the MutableCharSpan.
+     * @return Returns a CHIP_NO_ERROR if there was no error.
+     */
+    virtual CHIP_ERROR GetModeLabelByIndex(uint8_t modeIndex, chip::MutableCharSpan & label);
+
+    /**
+     * Get the mode value of the Nth mode in the list of modes.
+     * @param modeIndex The index of the mode to be returned. It is assumed that modes are indexable from 0 and with no gaps.
+     * @param value a reference to the uint8_t variable that is to contain the mode value.
+     * @return Returns a CHIP_NO_ERROR if there was no error.
+     */
+    virtual CHIP_ERROR GetModeValueByIndex(uint8_t modeIndex, uint8_t & value);
+
+    /**
+     * Get the mode tags of the Nth mode in the list of modes.
+     * The caller must make sure the List points to an existing buffer of sufficient size to hold the spec-required number
+     * of tags, and the size of the List is the size of the buffer.
```

Comment:
This seems dangerous. Is it possible for the caller to provide a size_t for the max number of items, so that size checks can be done?

Reply:
The caller in this case is the SDK, mostly. Basically what we're saying is that the caller has to provide space for the maximum allowed number of tags, which the spec defines as 8. If we add a size option, we will give a user to optionally pass in a smaller buffer which might then cause an "insufficient size" type of error which would be difficult to deal with on the device.

Reply:
What about if we add that the spec defines the maximum size as 8 in the doc?

Reply:
Tackled in a prevous comment.